designs/disruption-controls-by-method.md

```diff
@@ -0,0 +1,397 @@
+      // a 0s at the end.
+      // +kubebuilder:validation:Pattern=`^([0-9]+(m|h)+(0s)?)$`
+      // +kubebuilder:validation:Type="string"
+      // +optional
+      Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+}
+
+type DisruptionMethod string 
+
+const (
+	All DisruptionMethod = ""
+	Consolidation DisruptionMethod = "Consolidation" 
+	Drift DisruptionMethod = "Drift" 
+	Expiration DisruptionMethod = "Expiration" 
+	Emptiness DisruptionMethod = "Emptiness"
+)
+```
+
+
+##### Example
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    budgets:
+    - schedule: "* * * * *"
+      action: Drift
+      nodes: 10
+    - schedule: "* * * * *"
+      action: Drift
+      nodes: 10
+    # For all other actions , only allow 5 nodes to be disrupted at a time
+    - nodes: 5
+      schedule: "* * * * *"
+
+```
+
+In the original proposed spec, karpenter allows the user to specify up to [50 budgets](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/apis/v1beta1/nodepool.go#L96)
+
+If there are multiple active budgets, karpenter takes the most restrictive budget. This same principle will be applied to the disruption budgets in this approach. The only difference in behavior is that each window will either apply to a single action or all actions. 
+
+### Pros + Cons 
+* 👍 No nested definitions required 
+* 👍👍 Extending existing budgets api. No Breaking API Changes, completely backwards compatible  
+* 👎 With action being clearly tied to budgets, and other api logic being driven by disruption Method, we lose the chance to generalize per method controls 
+* 👎 Makes validation of a particular disruption 
+
+### Approach B: Defining Per Method Controls  
+Ideally, we could move all generic controls that easily map into other actions into one set of action controls, this applies to budgets and other various disruption controls that could be more generic. 
+### Proposed Spec 
+```go
+type Disruption struct {
+    All		  DisruptionSpec `json:defaults"`
+    Consolidation ConsolidationSpec `json:"consolidation"`
+    Drift         DriftSpec         `json:"drift"`
+    Expiration    ExpirationSpec    `json:"expiration"`
+    Emptiness     EmptinessSpec     `json:"emptiness"`
+}
+
+type DisruptionCommonSpec struct {
+    DisruptAfter string   `json:"disruptAfter"`
+    Budgets      []Budget `json:"budgets"`
+}
+
+type ConsolidationSpec struct {
+    DisruptionCommonSpec
+    ConsolidationPolicy string `json:"consolidationPolicy"`
+}
+
+type DriftSpec struct {
+    DisruptionCommonSpec
+}
+
+type ExpirationSpec struct {
+    DisruptionCommonSpec
+}
+
+type EmptinessSpec struct {
+    DisruptionCommonSpec
+}
+
+type Budget struct {
+    Nodes    string  `json:"nodes"`
+    Schedule *string `json:"schedule,omitempty"`
+    Duration *string `json:"duration,omitempty"`
+    Reason string 
+}
+}
+
+type DisruptionMethod string 
```

Comment:
I don't see how the `DisruptionMethod`(s) are used in approach B?

Reply:
DisruptionMethods aren't needing specification because you define the disruption method at the top level of the api.  We will however still need to use either these constants or the metrics constants https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/metrics/constants.go#L34

Reply:
These constants are used for internal representation mainly for the second approach